braidpool_spec.md

```diff
@@ -460,33 +675,73 @@ formed, fully signed and valid bitcoin transactions that can be broadcast. See
 [Payout Authorization](#payout-authorization) for considerations on how to
 sign/authorize this UHPO transaction.
 
+[kp]
+
+The UHPO transactions are signed and broadcast using the braidpool P2P
+network. In case of the pool shutting down, the miners will have to figure out a
+communication channel to broadcast their signature shares and reach agreement on
+spending the UHPO.
+
+This sounds like a problem.
+
+[kp]
+
 We don't ever want to actually have to broadcast this UHPO set transaction
 except in the case of pool failure. Similar to other optimistic protocols like
 Lightning, we will withhold this transaction from bitcoin and update it
-out-of-band with respect to bitcoin. With each new block we will update the UHPO
-set transaction to account for any new shares since the last block mined by the
-pool.
+out-of-band with respect to bitcoin. With each new block found by braidpool, we
+will update the UHPO set transaction to account for any new shares since the
+last block mined by the pool.
 
 Furthermore a decentralized mining pool should support "withdrawal" by hashers.
-This would take the form of a special message or transaction sent to the pool
+This will take the form of a special message or transaction sent to the pool
 (and agreed by consensus within the pool) to *remove* a hasher's output from the
 UHPO set transaction, and create a new separate transaction which pays that
 hasher, [authorizes](#payout-authorization) it, and broadcasts it to bitcoin.
 Miners may not cash out for the *current* difficulty adjustment window (because
 the share/BTC price is not yet decided), but may only cash out for the *last*
 (and older) difficulty adjustment window(s).
 
+[kp]
+
+We need to address potential race condition and instability periods when miners
+are cashing out and new bitcoin blocks are being found. If there is no such
+instability, we should explain why. I expect this question will be raised during
+review.
+
+
+[/kp]
+
+
 The share value for the *current* difficulty adjustment epoch is encoded
 proportionally in the UHPO transactions, however this is only for use in the
 case of catastrophic failure of Braidpool. During normal operation, the UHPO
 transaction is fixed at the end of the difficulty adjustment window when the
 share/BTC price for that epoch is known.
 
+[kp]
+
+Why is the share value fixed?
+
+We don't need a fixed share value. A miner's share value over a period can
+determined by their claimed difficulty. The difficulty is claimed at the outset,
+then all shares confirm the miner mined with that difficulty as some shares will
+get close to that difficulty.
+
+[/kp]
+
 # Payout Update and Settlement Signing
 
 Once a bitcoin block is mined by the pool, Braidpool will kick off a signing
 ceremony to create a new Payout Commitment and UHPO settlement transaction.
 
+[kp]
+
+How can this ceremony be disrupted during normal operation of braidpool, or by
+an attacker?
+
+[kp]
+
```

Comment:
If the attacker knows the IPs of miners in the ceremony he can DoS them.

Therefore IPs involved in the ceremony must be hidden from public. I suggest that when publishing a block, the miner finding the block encrypts the relevant IP/API endpoint in the corresponding bead for each other miner involved in the ceremony. Therefore only miners involved in the ceremony will know each other's IPs.

A second mechanism would be to use a mixnet or Tor or something. This question needs more thought.